bcda/models/models.go

```diff
@@ -40,14 +45,93 @@ func InitializeGormModels() *gorm.DB {
 type Job struct {
 	gorm.Model
 	ACO        ACO       `gorm:"foreignkey:ACOID;association_foreignkey:UUID"` // aco
-	ACOID      uuid.UUID `gorm:"primary_key; type:char(36)" json:"aco_id"`
+	ACOID      uuid.UUID `gorm:"type:char(36)" json:"aco_id"`
 	User       User      `gorm:"foreignkey:UserID;association_foreignkey:UUID"` // user
 	UserID     uuid.UUID `gorm:"type:char(36)"`
 	RequestURL string    `json:"request_url"` // request_url
 	Status     string    `json:"status"`      // status
+	JobCount   int
 	JobKeys    []JobKey
 }
 
+func (job *Job) CheckCompleted() (bool, error) {
+
+	// Trivial case, no need to keep going
+	if job.Status == "Completed" {
+		return true, nil
+	}
+	db := database.GetGORMDbConnection()
+	defer database.Close(db)
+
+	var completedJobs int64
+	db.Model(&JobKey{}).Where("job_id = ?", job.ID).Count(&completedJobs)
+
+	if int(completedJobs) >= job.JobCount {
+
+		staging := fmt.Sprintf("%s/%d", os.Getenv("FHIR_STAGING_DIR"), job.ID)
+		err := os.Remove(staging)
```

Comment:
`CheckCompleted` sounds like it's looking to see if the job is done, so it's not clear that there would be actions taken here like removing the staging directory. Could we rename to make that apparent, or make it a separate function?

Reply:
This should still be idempotent.  The staging dir it is deleting was housing all of the unencrypted BB files for just this job.  They should have all been moved at this point and this is just housekeeping.  It HAS to happen so I just wanted to make sure it was done in the easiest way to not mess up.

Reply:
Understood, and idempotence isn't in question. Functions shouldn't have side effects. Could the staging directory deletion be a separate function, called when this one returns `true`? That would also make it possible to test the behavior of job completion and file cleanup separately.

Reply:
Returning True just means that the job is completed, it could have been completed previously or it could be marked as completed just now.  

This is just a task that 100% has to happen when the job is finally marked as complete.  If I split it up I am concerned that the next time someone uses it they won't know about this requirement and will choose not to do it.  By keeping it here they can't make that mistake.  The dir it is deleting is scoped only to this job and won't affect any other jobs.

The way I see it if I close my front door I need to lock it.  It isn't two things, it is one thing close the door and lock it.  A function called CloseDoor() ends in a closed and locked door and I can't end up leaving the door unlocked.

I discussed with Ryan in my 1 on 1 today and he seemed fine with it.

Reply:
I looked in my copy of "Clean Code" because I think we are both directly arguing for points he makes in that book and this is a real judgement call.  Chapter 3 (Functions) says functions should "Do One Thing" and "Have No Side Effects".

> So, another way to know that a function is doing more than 'one thing' is if you can extract another function from it with a name that is not merely a restatement of its implementation.

  I would state that the abstraction of checkComplete()  is this:
`Check if a job is in a completed status and if it was not previously in a completed status clean up the job's staging directory.`
Any further refactoring would, in my opinion, unnecessarily complicate things and force the caller of the function to take important steps that might be missed

>Side Effects are lies.  Your function promises to do one thing, but it also does other hidden things.  Sometimes it will make unexpected changes to the variables of its own class.  Sometimes it will make them to parameters passed into the function or to system globals.  In either case they are devious and damaging mistruths that often result in strange temporal couplings and order dependencies

I feel that the side effects of this function are sufficiently tied to the job in question that they do not have any global effects and will not cause any temporal couplings or dependencies.  The creation of a job key only happens after the file in question has been encrypted and moved out of the staging directory.  As such, when this function decides to delete the staging directory it should be deleting an empty directory and if it isn't, we have solved a problem (storing unencrypted BB data) that we didn't know we had in the first place.

Reply:
My other suggestion was to rename the function. It sounds like this logic is all closely tied together, so can we call it something that will more clearly imply what it does?

Reply:
renamed to CheckJobCompletedAndCleanup